src/ray/util/process.cc

```diff
@@ -629,9 +629,15 @@ bool IsProcessAlive(pid_t pid) {
   }
   return false;
 #else
+  // Check pid is alive or not
   if (kill(pid, 0) == -1 && errno == ESRCH) {
     return false;
   }
+  // If it's a child, check whether it's zombie
+  int ret = waitpid(pid, nullptr, WNOHANG);
+  if (ret == pid) {
```

Comment:
This is the subtle part. if ret == -1, it doesn't mean that the pid has exited.
It only means there are errors.

One example is that, core worker also use this to check raylet's healthiness, and raylet is not a child of core worker, so it'll fail with ret == -1.

Reply:
Hmm doesn't need to be done in this PR, but something we could do in the med term is (maybe after migrating to boost);

1. Have some timeout mechanism assuming most of issues are transient
2. I believe we can detect the case you are talking about from the error code, so we can handle it differently.

Reply:
I did research of that since I did what you mentioned before and it failed some tests.
In the end I found this post and it said it's not possible with pure posix api unless it's parent.

https://stackoverflow.com/questions/1619020/how-to-check-if-a-process-with-a-pid-x-is-a-zombie